torch/_dynamo/variables/tensor.py

```diff
@@ -677,7 +682,59 @@ def redistribute_fn_with_prim_types(x):
                 # Intermediary
-                unimplemented("Intermediary tensors with registered hooks - NYI")
+
+                # This function is only invoked if the passed-in func has no source,
+                # useful for intermediary created functions.
+                def _lifted_fn_src(name):
+                    base = name
+                    for i in itertools.count():
+                        if name not in tx.output.global_scope:
+                            src = GlobalSource(name)
+                            tx.output.guards.add(
+                                src.make_guard(GuardBuilder.WEAKREF_ALIVE)
+                            )
+                            tx.output.install_global(name, weakref.ref(fn))
+                            break
+                        name = f"{base}_{i}"
+                    return src
+
+                # Either we have a source already, or we must make a weakref to the original fn
+                # and reference that.
+                src = fn_var.source if fn_var.source else _lifted_fn_src(name)
+
+                from .builder import GraphArg
+
+                # This wraps our user provided fn with a function that intercedes and
+                # uses our `invoke` higher order op to record a hook invocation in bwd graph.
+                fn = functools.partial(trace_wrapped, fn=fn)
+
+                # This little piece of logic ensures we only ever lift a given hook up once
+                # this important, because a dynamo->aot_autograd invariant is no duplicate sources in inputs.
+                hook_proxy = None
+                for node in tx.output.root_tracer.graph.nodes:
+                    if node.op == "placeholder":
+                        if "source" in node.meta:
+                            if node.meta["source"] == src:
+                                hook_proxy = node.meta["proxy"]
+                                break
+
+                if hook_proxy is None:
+                    hook_proxy = tx.output.root_tracer.create_graph_input(
+                        name, type(fn), source=src
+                    )
+                    hook_proxy.node.meta["source"] = src
+                    grapharg = GraphArg(src, fn, False, None)
+                    hook_proxy.node.meta["grapharg"] = grapharg
+                    hook_proxy.node.meta["proxy"] = hook_proxy
+                # this is a stepping stone implementation for now, the real POR to avoid recompiling on hook identity
+                # is to add an op in forward that is persisted through functionalization, and stashes hooks in a well defined
+                # place - this allows relaxing specialization from hook identity to # of hooks (and their positions)
+                self.as_proxy().register_hook(hook_proxy)
+
+                if fn_var.source:
+                    tx.output.guards.add(
+                        fn_var.source.make_guard(GuardBuilder.ID_MATCH)
+                    )
```

Comment:
This seems... incomplete? If the fn_var doesn't have a source, you still need to add guards no?